refactor: [#100] move high-level dvc library methods

josecelano opened the pull request:
These methods should not be in the generic DVC mod.
They filter files that belong to a library (media files).
The generic DVC mod should know nothing about librarian libraries and
`namecodes` specification.

josecelano commented:
I've only moved some methods so far. I want to refactor a little bit:

- [ ] Change [dvc_dif_parser](https://github.com/Nautilus-Cyberneering/nautilus-librarian/pull/105/files#diff-54496ff68b030a5d2a7a859b719d8583dcd132cbe07d684bff37d4a29209a765) in domain to make it similar to the [dvc_diff_parser](https://github.com/Nautilus-Cyberneering/nautilus-librarian/blob/main/src/nautilus_librarian/mods/dvc/domain/dvc_diff_parser.py) in the mod. The `dvc_diff_parser` in the domain always returns only media files, filtering non-media files.

josecelano commented:
I've created new classes to wrap the strings representing filenames, paths, directories. I'm planning to use them in `DvcDiffParser` and `DvcDiffMediaParser`. I will probably create a new `filesystem` mod because the `dvc` mod can also use them. I also expect other mods to sue them as they are very basic blocks for CI/CD tasks.

josecelano commented:
TODO: 

- [ ] use `Filename`, `Filepath` and `Directory` classes on `DvcDiffParser` and `DvcDiffMediaParser`.
- [ ] I'm likely to introduce also a class `RenamedFile` representing the `dvc diff` renamed file.

josecelano commented:
hi @da2ce7 @yeraydavidrodriguez I have finished with the main refactor but there is still one test failing.

@yeraydavidrodriguez could you take a look at it, please?

This is the test:

```s
poetry run pytest -vvv -s -k "given_a_dvc_diff_object_with_a_renamed_gold_image_it_should_commit_the_renamed_base_image_to_git"
```

For some reason the final result is OK but commits are not the way we expect them to be in the test:

![image](https://user-images.githubusercontent.com/58816/159042084-d04e7587-a91f-41d0-b687-d6fd05b16707.png)

It seems the "move Base image" commit has been split into 2 commits: one to delete the only file and one to add the new one, instead of having only one commit with both changes.

![image](https://user-images.githubusercontent.com/58816/159042435-e82201fb-6122-4e61-8bcc-3066b2be7052.png)

Maybe that was your intention but then:

1. I do not know why the test was working before the refactor.
2. We should change the commit messages.

I do not know the reason why this is happening but maybe when you change the action to use the "dvc move" to rename the Base image you forget the change the commits messages.

josecelano commented:
@yeraydavidrodriguez I can confirm that it's working on the `main` branch:

![image](https://user-images.githubusercontent.com/58816/159043938-91b9ba18-4488-4f38-b309-a99350317846.png)

so it must be something I've changed in this branch. Anyway, it would be nice if you can take a look at it with fresh eyes.

yeraydavidrodriguez commented:
Of course @josecelano , let me check it carefully.

yeraydavidrodriguez commented on src/nautilus_librarian/mods/dvc/domain/diff/renamed_path.py:
```diff
@@ -0,0 +1,23 @@
+        self.old_path = RelativeFilepath(old_path)
+
+    def __str__(self) -> str:
+        return str(self.path)
```
It seems strange to me that the string conversion conceals the fact that is a renaming path removing the old path. In my opinion, this can produce bugs difficult to locate.

yeraydavidrodriguez commented on tests/test_nautilus_librarian/test_typer/test_commands/test_workflows/test_actions/test_validate_filenames_action.py:
```diff
@@ -17,29 +17,12 @@ def given_a_dvc_diff_object_it_should_validate_the_filename_of_the_new_media_fil
     assert result.contains_text("000001-32.600.2.tif ✓")
 
 
-def given_a_dvc_diff_object_it_should_validate_the_filename_of_the_modified_media_files():
```
Why was this test removed?

josecelano replied:
I assumed when a file is only modified only the contents change, so the filename does not change and we do not need to validate it again, right?

josecelano commented on tests/test_nautilus_librarian/test_mods/test_dvc/test_domain/test_diff/test_renamed_path.py:
```diff
@@ -0,0 +1,9 @@
+def it_should_contain_the_old_and_new_relative_paths_for_a_renamed_file():
+
+    assert (
+        str(Path("data/000001/32/000001-32.600.2.tif"))
```
I've changed it

josecelano commented:
hi @yeraydavidrodriguez I found the problem. This function was also returning renamed files:

```
def extract_added_files_from_dvc_diff(dvc_diff_json):
    dvc_diff = Parser.from_json(dvc_diff_json)
    return dvc_diff.filter(
        exclude_deleted=True, exclude_modified=True, exclude_renamed=True
    ).as_plain_list()
```

So the renamed files were committed as new files. And the next step (committing renamed files) was only committing the deleted files because the new one were already committed.

Besides, the test for `extract_added_files_from_dvc_diff` was working because I was only adding "added" files in the dvc diff. I've been changing tests to use `dummy_full_dvc_diff` with at least one item of each type (added, modified, renamed, deleted) to make sure not only that we get want we want but also that we do not get what we do not want.
